File: src/tools/implementations/todo.ts
/**
 * Todo Tools - Conversation-scoped todo list management for agents
 *
 * Provides todo_write functionality for full state replacement.
 * Todos are stored on the Conversation object and persisted with it.
 */

import type { ConversationToolContext } from "@/tools/types";
import type { ConversationStore } from "@/conversations/ConversationStore";
import type { TodoItem, TodoStatus } from "@/services/ral/types";
import type { AISdkTool } from "@/tools/types";
import { tool } from "ai";
import { z } from "zod";

// ============================================================================
// Helper functions
// ============================================================================

function getCurrentTimestamp(): number {
    return Date.now();
}

/**
 * Simple deterministic hash for generating fallback IDs.
 * Returns a hex string based on the input.
 */
function simpleHash(str: string): string {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
        const char = str.charCodeAt(i);
        hash = ((hash << 5) - hash) + char;
        hash = hash & hash; // Convert to 32bit integer
    }
    // Convert to positive hex string
    return Math.abs(hash).toString(16);
}

/**
 * Generates a slug-style ID from a title.
 * Converts to lowercase, replaces spaces/special chars with hyphens, and removes consecutive hyphens.
 * Falls back to a deterministic hash if the title produces an empty slug (e.g., emoji-only or non-ASCII titles).
 */
function generateIdFromTitle(title: string): string {
    const slug = title
        .toLowerCase()
        .replace(/[^a-z0-9\s-]/g, "") // Remove special characters except spaces and hyphens
        .replace(/\s+/g, "-") // Replace spaces with hyphens
        .replace(/-+/g, "-") // Replace consecutive hyphens with single hyphen
        .replace(/^-|-$/g, ""); // Remove leading/trailing hyphens

    // Fall back to hash if slug is empty (non-ASCII/emoji-only titles)
    if (slug === "") {
        return `todo-${simpleHash(title)}`;
    }

    return slug;
}

/**
 * Validates and writes the complete todo list, replacing all existing items.
 * Implements safety check to prevent accidental deletions.
 */
function writeTodosToConversation(
    conversation: ConversationStore,
    agentPubkey: string,
    newItems: Array<{
        id: string;
        title: string;
        description?: string; // undefined means "preserve existing" for updates
        status: TodoStatus;
        skip_reason?: string;
    }>,
    force: boolean
): {
    success: boolean;
    items: TodoItem[];
    error?: string;
    missingIds?: string[];
} {
    const existingTodos = conversation.getTodos(agentPubkey);
    const now = getCurrentTimestamp();

    // Build a map of new item IDs for quick lookup
    const newItemIds = new Set(newItems.map((item) => item.id));

    // Check for duplicate IDs in the input array
    if (newItemIds.size !== newItems.length) {
        const seen = new Set<string>();
        const duplicates: string[] = [];
        for (const item of newItems) {
            if (seen.has(item.id)) {
                duplicates.push(item.id);
            }
            seen.add(item.id);
        }
        return {
            success: false,
            items: [],
            error: `Duplicate IDs in input: ${duplicates.join(", ")}`,
        };
    }

    // Safety check: find any existing IDs that are missing from the new list
    const existingIds = existingTodos.map((t) => t.id);
    const missingIds = existingIds.filter((id) => !newItemIds.has(id));

    if (missingIds.length > 0 && !force) {
        return {
            success: false,
            items: [],
            error: `Safety check failed: ${missingIds.length} existing item(s) would be removed. Use force=true to allow removal.`,
            missingIds,
        };
    }

    // Validate skip_reason requirement
    for (const item of newItems) {
        if (item.status === "skipped" && !item.skip_reason) {
            return {
                success: false,
                items: [],
                error: `Validation failed: skip_reason is required when status='skipped' (item: ${item.id})`,
            };
        }
    }

    // Build the new todo list, preserving timestamps for existing items
    // Array order determines position (index-based ordering)
    const existingMap = new Map(existingTodos.map((t) => [t.id, t]));
    const newTodos: TodoItem[] = newItems.map((item) => {
        const existing = existingMap.get(item.id);
        return {
            id: item.id,
            title: item.title,
            // Preserve existing description if not provided in update (undefined means "keep existing")
            description: item.description ?? existing?.description ?? "",
            status: item.status,
            skipReason: item.skip_reason,
            createdAt: existing?.createdAt ?? now,
            updatedAt: existing && existing.status === item.status ? existing.updatedAt : now,
        };
    });

    // Persist the new list
    conversation.setTodos(agentPubkey, newTodos);

    return {
        success: true,
        items: newTodos,
    };
}

// ============================================================================
// todo_write
// ============================================================================

const todoWriteItemSchema = z.object({
    id: z.string().optional().describe("Optional unique identifier. Auto-generated from title if not provided."),
    title: z.string().describe("Short human-readable title for the todo item"),
    description: z.string().optional().describe("Optional detailed description. When updating existing items, omitting preserves the current description. For new items, defaults to empty string."),
    status: z
        .enum(["pending", "in_progress", "done", "skipped"])
        .describe("Current status of the item"),
    skip_reason: z
        .string()
        .optional()
        .describe("Required when status='skipped' - explain why this item was skipped"),
});

const todoWriteSchema = z.object({
    todos: z.array(todoWriteItemSchema).describe("The complete todo list. All items must be provided - this replaces the entire list."),
    force: z
        .boolean()
        .optional()
        .default(false)
        .describe("When true, allows removing items from the list. Default false (safety check prevents removals)."),
});

type TodoWriteInput = z.infer<typeof todoWriteSchema>;

interface TodoWriteOutput {
    success: boolean;
    message: string;
    totalItems: number;
    error?: string;
    missingIds?: string[];
}

async function executeTodoWrite(
    input: TodoWriteInput,
    context: ConversationToolContext
): Promise<TodoWriteOutput> {
    const conversation = context.getConversation();

    const result = writeTodosToConversation(
        conversation,
        context.agent.pubkey,
        input.todos.map((item) => ({
            id: item.id ?? generateIdFromTitle(item.title),
            title: item.title,
            description: item.description, // Pass undefined to preserve existing; handled in writeTodosToConversation
            status: item.status as TodoStatus,
            skip_reason: item.skip_reason,
        })),
        input.force ?? false
    );

    // Generate a concise message based on the operation
    let message: string;
    if (!result.success) {
        message = result.error || "Failed to write todos";
    } else if (result.items.length === 0) {
        message = "Todo list cleared";
    } else if (result.items.length === 1) {
        message = `Todo list updated with 1 item`;
    } else {
        message = `Todo list updated with ${result.items.length} items`;
    }

    return {
        success: result.success,
        message,
        totalItems: result.items.length,
        error: result.error,
        missingIds: result.missingIds,
    };
}

export function createTodoWriteTool(context: ConversationToolContext): AISdkTool {
    const aiTool = tool({
        description:
            "Write the complete todo list, replacing all existing items. Provide ALL items you want to exist - " +
            "this is a full state replacement. By default, removing existing items is blocked (safety check). " +
            "Set force=true to allow item removal. Each item requires: title and status. " +
            "Optional: id (auto-generated from title if not provided), description (preserved on update if omitted, empty for new items). " +
            "Use skip_reason when status='skipped'. " +
            "NOTE: If an existing item used a custom ID and you omit the id field, the auto-generated ID won't match, " +
            "which will trigger the safety check (or require force=true). Always include IDs for items with custom IDs.",
        inputSchema: todoWriteSchema,
        execute: async (input: TodoWriteInput) => {
            return await executeTodoWrite(input, context);
        },
    });

    Object.defineProperty(aiTool, "getHumanReadableContent", {
        value: ({ todos }: TodoWriteInput) => {
            if (todos.length === 0) {
                return "Clearing todo list";
            }
            if (todos.length === 1) {
                return `Writing todo list with 1 item: "${todos[0].title}"`;
            }
            return `Writing todo list with ${todos.length} items`;
        },
        enumerable: false,
        configurable: true,
    });

    return aiTool as AISdkTool;
}

File: crates/tenex-agent/src/tools/todo.rs
use rig::{completion::ToolDefinition, tool::Tool};
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::{
    sync::{Arc, Mutex},
    time::{SystemTime, UNIX_EPOCH},
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TodoItem {
    pub id: String,
    pub title: String,
    pub description: String,
    pub status: TodoStatus,
    pub skip_reason: Option<String>,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum TodoStatus {
    Pending,
    InProgress,
    Done,
    Skipped,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct TodoWriteItem {
    pub id: Option<String>,
    pub title: String,
    pub description: Option<String>,
    pub status: TodoStatus,
    pub skip_reason: Option<String>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct TodoWriteArgs {
    pub todos: Vec<TodoWriteItem>,
    pub force: Option<bool>,
}

#[derive(Debug, thiserror::Error)]
#[error("{0}")]
pub struct TodoError(String);

pub struct TodoWriteTool {
    todos: Arc<Mutex<Vec<TodoItem>>>,
}

impl TodoWriteTool {
    pub fn new(todos: Arc<Mutex<Vec<TodoItem>>>) -> Self {
        Self { todos }
    }
}

fn slug_from_title(title: &str) -> String {
    let slug: String = title
        .to_lowercase()
        .chars()
        .map(|c| if c.is_alphanumeric() { c } else { '-' })
        .collect::<String>()
        .split('-')
        .filter(|s| !s.is_empty())
        .collect::<Vec<_>>()
        .join("-");
    if slug.is_empty() {
        format!("todo-{:x}", title.len())
    } else {
        slug
    }
}

fn now_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_or(0, |d| d.as_millis() as i64)
}

pub fn format_todos_reminder(todos: &[TodoItem]) -> String {
    if todos.is_empty() {
        return String::new();
    }

    let pending = todos
        .iter()
        .filter(|t| t.status == TodoStatus::Pending)
        .count();
    let in_progress = todos
        .iter()
        .filter(|t| t.status == TodoStatus::InProgress)
        .count();
    let done = todos
        .iter()
        .filter(|t| t.status == TodoStatus::Done)
        .count();
    let skipped = todos
        .iter()
        .filter(|t| t.status == TodoStatus::Skipped)
        .count();

    let mut lines = vec![
        "<system-reminder>".to_string(),
        "<agent-todos>".to_string(),
        String::new(),
        format!(
            "Status: {} pending, {} in progress, {} done, {} skipped",
            pending, in_progress, done, skipped
        ),
        String::new(),
    ];

    for t in todos {
        let marker = match t.status {
            TodoStatus::Pending => "[ ]",
            TodoStatus::InProgress => "[~]",
            TodoStatus::Done => "[x]",
            TodoStatus::Skipped => "[-]",
        };
        lines.push(format!("{} {} (id: {})", marker, t.title, t.id));
    }

    lines.push(String::new());
    lines.push(
        "Use `todo_write` to update statuses. Mark items in_progress when starting, done when complete."
            .to_string(),
    );

    if pending > 0 {
        lines.push(String::new());
        lines.push(format!(
            "**ATTENTION:** You have {pending} pending todo item(s) that need to be addressed."
        ));
    }

    lines.push("</agent-todos>".to_string());
    lines.push("</system-reminder>".to_string());

    lines.join("\n")
}

/// Apply a `todo_write` invocation to the shared todo list. Encapsulates
/// validation (skip_reason, duplicate IDs, removal protection) and the
/// preservation of `created_at` / `description` for items that already
/// existed. Used by both the `todo_write` tool and `run_workflow`, which
/// activates a freshly-generated checklist with the same semantics.
pub fn apply_todo_write(
    todos: &Arc<Mutex<Vec<TodoItem>>>,
    args: TodoWriteArgs,
) -> Result<String, TodoError> {
    let force = args.force.unwrap_or(false);
    let now = now_ms();

    // Validate skip_reason presence
    for item in &args.todos {
        if item.status == TodoStatus::Skipped && item.skip_reason.is_none() {
            return Ok(format!(
                "Error: skip_reason is required when status='skipped' (item: {:?})",
                item.id.as_deref().unwrap_or(&item.title)
            ));
        }
    }

    // Validate no duplicate IDs in input
    let mut seen_ids = std::collections::HashSet::new();
    for item in &args.todos {
        let id = item
            .id
            .clone()
            .unwrap_or_else(|| slug_from_title(&item.title));
        if !seen_ids.insert(id.clone()) {
            return Ok(format!("Error: duplicate id '{id}' in input"));
        }
    }

    let mut todos = todos
        .lock()
        .map_err(|_| TodoError("Failed to acquire todo lock".to_string()))?;

    // Safety check: detect removals
    if !force {
        let new_ids: std::collections::HashSet<String> = args
            .todos
            .iter()
            .map(|t| t.id.clone().unwrap_or_else(|| slug_from_title(&t.title)))
            .collect();
        let missing: Vec<&str> = todos
            .iter()
            .filter(|t| !new_ids.contains(&t.id))
            .map(|t| t.id.as_str())
            .collect();
        if !missing.is_empty() {
            return Ok(format!(
                "Error: {} existing item(s) would be removed: {}. Use force=true to allow.",
                missing.len(),
                missing.join(", ")
            ));
        }
    }

    // Build new list, preserving created_at/description from existing items
    let new_todos: Vec<TodoItem> = args
        .todos
        .into_iter()
        .map(|item| {
            let id = item.id.unwrap_or_else(|| slug_from_title(&item.title));
            let existing = todos.iter().find(|t| t.id == id);
            let created_at = existing.map_or(now, |e| e.created_at);
            let description = item
                .description
                .or_else(|| existing.map(|e| e.description.clone()))
                .unwrap_or_default();
            let status_changed = existing.is_none_or(|e| e.status != item.status);
            let updated_at = if status_changed {
                now
            } else {
                existing.map_or(now, |e| e.updated_at)
            };
            TodoItem {
                id,
                title: item.title,
                description,
                status: item.status,
                skip_reason: item.skip_reason,
                created_at,
                updated_at,
            }
        })
        .collect();

    let count = new_todos.len();
    *todos = new_todos;

    let summary: Vec<String> = todos
        .iter()
        .map(|t| {
            let status_icon = match t.status {
                TodoStatus::Pending => "○",
                TodoStatus::InProgress => "◉",
                TodoStatus::Done => "✓",
                TodoStatus::Skipped => "⊘",
            };
            format!("{status_icon} [{}] {}", t.id, t.title)
        })
        .collect();

    Ok(format!(
        "Todo list updated ({count} items):\n{}",
        summary.join("\n")
    ))
}

impl Tool for TodoWriteTool {
    const NAME: &'static str = "todo_write";
    type Error = TodoError;
    type Args = TodoWriteArgs;
    type Output = String;

    async fn definition(&self, _prompt: String) -> ToolDefinition {
        ToolDefinition {
            name: Self::NAME.to_string(),
            description: "Write the complete todo list, replacing all existing items. \
                Provide ALL items you want to exist — this is a full state replacement. \
                By default, removing existing items is blocked (use force=true to allow). \
                Use skip_reason when status='skipped'."
                .to_string(),
            parameters: json!({
                "type": "object",
                "properties": {
                    "todos": {
                        "type": "array",
                        "description": "Complete todo list (replaces all existing items)",
                        "items": {
                            "type": "object",
                            "properties": {
                                "id": { "type": "string", "description": "Unique ID (auto-generated from title if omitted)" },
                                "title": { "type": "string", "description": "Short description" },
                                "description": { "type": "string", "description": "Detailed description (omit to preserve existing value on updates)" },
                                "status": {
                                    "type": "string",
                                    "enum": ["pending", "in_progress", "done", "skipped"]
                                },
                                "skip_reason": { "type": "string", "description": "Required when status=skipped" }
                            },
                            "required": ["title", "status"]
                        }
                    },
                    "force": {
                        "type": "boolean",
                        "description": "Allow removing existing items (default: false)"
                    }
                },
                "required": ["todos"]
            }),
        }
    }

    async fn call(&self, args: TodoWriteArgs) -> Result<Self::Output, TodoError> {
        apply_todo_write(&self.todos, args)
    }
}
